Drop guava

dmandalidis opened the pull request.

dmandalidis committed:
Drop guava (fixes #74)

- Remove workaround for JERSEY-2878

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -969,7 +967,7 @@ public TopResults topContainer(final String containerId, final String psArgs)
     try {
       WebTarget resource = resource().path("containers").path(containerId).path("top");
-      if (!Strings.isNullOrEmpty(psArgs)) {
+      if (psArgs != null && "".equals(psArgs.trim())) {
```
Shouldn't this be checking that psArgs is NOT empty?

dmandalidis replied:
good catch

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -2874,12 +2872,12 @@ public static Builder builder() {
+    	.filter(cert -> cert != null)
+    	.findFirst()
+    	.orElseThrow(() -> new NoSuchElementException("Cannot find docker certificated path")));
+    
```
nit: whitespace

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -2874,12 +2872,12 @@ public static Builder builder() {
+    	.stream()
+    	.filter(cert -> cert != null)
+    	.findFirst()
+    	.orElseThrow(() -> new NoSuchElementException("Cannot find docker certificated path")));
```
I'm not very savvy with stream syntax, but this looks mostly reasonable.  The orElseThrow makes sense, but how does the Exception get handled when it occurs?

dmandalidis replied:
The same way as before; the replaced `Iterables.find` throws a `NoSuchElementException` if the element is not found

tony-- commented on src/main/java/org/mandas/docker/client/DefaultLogStream.java:
```diff
@@ -23,16 +23,14 @@
 import java.nio.ByteBuffer;
 
+import com.google.common.collect.AbstractIterator;
+import com.google.common.io.Closer;
```
Is this still a work in progress?  If not, were these missed?

dmandalidis replied:
done

tony-- commented on src/main/java/org/mandas/docker/client/DockerHost.java:
```diff
@@ -23,14 +23,13 @@
 
+import com.google.common.base.MoreObjects;
+import com.google.common.net.HostAndPort;
+
```
More guava remnants

tony-- commented:
I like this idea - it's something my team has talked about but never explored.  I scanned through the changes quickly and it looks like this is still a work in progress?

dmandalidis commented:
Hi @tony-- 

I 'm done with this. Please ping me if you 're going to review it so I can wait a little bit before merging it

tony-- commented:
@dmandalidis - Thanks for the effort on this.  I'll take a look ASAP, but I might not get to it until Tuesday (US West Coast, +10 hours your time, I think)

dmandalidis commented:
no problem, take yr time

tony-- commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@
+* `EventStream#close` and `LogStream#close` can now throw `IOException`
+* `LogStream#attach(OutputStream, OutputStream, boolean)` method removed. Methods should
+not close streams they do not manage.
+* `LogStream#attach(OutputStream, OutputStream)` can now throw `IOException`
```
Consider adding a bullet for "removed dependency on Guava, which resulted in some of the breaking changes"

dmandalidis replied:
This would be added along with the commits just before releasing 3.0.0. I 've added these bullets here so I won't forget them :)

tony-- commented on src/main/java/org/mandas/docker/client/AnsiProgressHandler.java:
```diff
@@ -46,7 +45,7 @@ public AnsiProgressHandler() {
   public AnsiProgressHandler(PrintStream out) {
     this.out = out;
-    idsToLines = Maps.newHashMap();
+    idsToLines = new HashMap<>();
```
~~seems like this should be `idsToLines = new HashMap<String, Integer>();` in order to avoid a warning?~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

tony-- commented on src/main/java/org/mandas/docker/client/CompressedDirectory.java:
```diff
@@ -137,14 +136,14 @@ public void close() throws IOException {
+  static List<DockerIgnorePathMatcher> parseDockerIgnore(Path dockerIgnorePath)
       throws IOException {
-    final ImmutableList.Builder<DockerIgnorePathMatcher> matchersBuilder = ImmutableList.builder();
+    final List<DockerIgnorePathMatcher> matchersBuilder = new ArrayList<>();
```
~~Should `new ArrayList<>` be typed as `DockerIgnorePathMatcher` in order to avoid a warning?~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

tony-- commented on src/main/java/org/mandas/docker/client/CompressedDirectory.java:
```diff
@@ -253,15 +251,16 @@ private static String getNotSeparatorPattern(String separator) {
       // .dockerignore matchers need to be read from the bottom of the file, 
       // so the given list should be reversed before using it.
-      this.ignoreMatchers = ignoreMatchers.reverse();
+      this.ignoreMatchers = new ArrayList<>(ignoreMatchers);
```
~~should new ArrayList<> be typed?~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

tony-- replied:
For consistency, should this be an unmodifiableList?

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -264,7 +259,7 @@ private LoadProgressHandler(ProgressHandler delegate) {
 
     private Set<String> getImageNames() {
-      return ImmutableSet.copyOf(imageNames);
+      return unmodifiableSet(new HashSet<>(imageNames));
```
~~Should this be typed?~~
~~I'm afraid all of these "should this be typed?" comments are a waste, but I'll keep calling out in case you want to type them.~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -431,8 +427,9 @@ public DefaultDockerClient(final URI uri, final DockerCertificatesStore dockerCe
   protected DefaultDockerClient(final Builder builder) {
-    final URI originalUri = checkNotNull(builder.uri, "uri");
-    checkNotNull(originalUri.getScheme(), "url has null scheme");
+	requireNonNull(builder.uri, "uri");
```
nit: indenting

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -536,7 +534,7 @@ private String toRegExp(String hostnameWithWildcards) {
   @Override
   public String getHost() {
-    return fromNullable(uri.getHost()).or("localhost");
+    return ofNullable(uri.getHost()).orElse("localhost");
```
I hadn't used this before and had to read the docs.  Seems pretty cool!  Let's see if I understand.
if `uri.getHost()` is non-null, `ofNullable` returns an Optional containing the value of uri.getHost().
if `uri.getHost()` is null, `ofNullable` returns an empty Optional.
if the invoking Optional has a value, `orElse` returns that value.
if the invoking Optional does not have a value, `orElse` returns "localhost".
Is that right?

dmandalidis replied:
correct

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -2737,7 +2737,7 @@ private void requestAndTail(final String method, final ProgressHandler handler,
     if (response.getStatusInfo().getFamily() != Response.Status.Family.SUCCESSFUL) {
       throw new DockerRequestException(method, resource.getUri(), response.getStatus(),
-                message(response), null);
+    		  response.readEntity(String.class), null);
```
Seems like a nice change here.

tony-- commented on src/main/java/org/mandas/docker/client/DefaultDockerClient.java:
```diff
@@ -2874,12 +2859,12 @@ public static Builder builder() {
+    	.filter(cert -> cert != null)
+    	.findFirst()
+    	.orElseThrow(() -> new NoSuchElementException("Cannot find docker certificated path")));
+    
```
This was marked resolved but it looks like the whitespace is still there?

tony-- commented on src/main/java/org/mandas/docker/client/DockerHost.java:
```diff
@@ -81,12 +72,17 @@ private DockerHost(final String endpoint, final String certPath) {
-
-      this.port = hostAndPort.getPortOrDefault(defaultPort());
-      this.address = isNullOrEmpty(hostText) ? DEFAULT_ADDRESS : hostText;
+      final String scheme = certPath == null || "".equals(certPath.trim()) ? "http" : "https";
```
It might be possible for certPath to be non-null but not exist.  Would it make sense to check if certPath.exists() and has certs in it?  I'm not sure, just thought of it after comparing with the similar code [here](https://github.com/dmandalidis/docker-client/pull/78/files#diff-497f14a9cf87565d6db0dab39d4d5a47R2879)

dmandalidis replied:
The `DockerHost#from(String, String)` is publicly exposed so technically the `certPath` can be anything. 

In general, I really didn't question the existing codebase to avoid mixing additional refactorings which would make this PR unreviewable. My gut feeling is that we 're checking too much stuff and we 're doing extensive checks just for protecting the clients sending us nonsense. I don't remember where, but I saw it multiple times resulting in correlation problems between what you 've passed and the error we print after sanitization. We could do the check you propose and fail early but I believe this is not in scope for this PR as well as addressing the code duplication between this method and the similar one in `DefaultDockerClient`

tony-- replied:
Ok, sounds good, thanks for for explaining.

tony-- commented on src/main/java/org/mandas/docker/client/DockerHost.java:
```diff
@@ -243,14 +242,14 @@ static String configPathFromEnv() {
+		+ ",bindUri=" + bindUri
+		+ ",address=" + address
+		+ ",port=" + port
+		+ ",certPath=" + certPath == null? "null": certPath
```
if you decide to check for existence above, you should check for existence here also

tony-- commented on src/main/java/org/mandas/docker/client/EventStream.java:
```diff
@@ -20,48 +20,51 @@
+	} catch (IOException e) {
+		throw new RuntimeException(e);
+	}
+	return next != null;
```
nit: indent is 4 instead of 2

tony-- commented on src/main/java/org/mandas/docker/client/EventStream.java:
```diff
@@ -20,48 +20,51 @@
+	}
+	Event value = next;
+	next = null;
+	return value;
```
nit: indent is 4 instead of 2

tony-- commented on src/main/java/org/mandas/docker/client/ImageRef.java:
```diff
@@ -87,15 +84,14 @@ public String getRegistryUrl() {
+		+ "registry=" + registry != null? registry: null
+		+ ",image=" + image != null? image: null
+		+ ",tag=" + tag != null? tag: null
+		+ "}";
```
nit: indent is 4 instead of 2

tony-- commented on src/main/java/org/mandas/docker/client/LogMessage.java:
```diff
@@ -34,8 +34,10 @@ public LogMessage(final int streamId, final ByteBuffer content) {
-    this.stream = checkNotNull(stream, "stream");
-    this.content = checkNotNull(content, "content");
+	requireNonNull(stream, "stream");
+	requireNonNull(content, "content");
```
nit: indent is 4 instead of 2

dmandalidis replied:
it's not (seems like a github UI thing)

tony-- commented on src/main/java/org/mandas/docker/client/LogReader.java:
```diff
@@ -66,15 +63,31 @@ public LogMessage nextMessage() throws IOException {
   }
 
+  private int readAllBytes(byte[] buf) throws IOException {
+	int total = 0;
+    while (total < buf.length) {
+      int result = stream.read(buf, total, buf.length - total);
+      if (result == -1) {
+        break;
+      }
+      total += result;
+    }
+    
+    return total;
+  }
```
nit: indent is 4 instead of 2

tony-- commented on src/main/java/org/mandas/docker/client/LogReader.java:
```diff
@@ -66,15 +63,31 @@ public LogMessage nextMessage() throws IOException {
-    // We cannot call the stream's close method because it an instance of UncloseableInputStream,
-    // where close is a no-op.
-    copy(stream, nullOutputStream());
+	stream.close();  
```
nit: indent is 4 instead of 2

dmandalidis replied:
same github UI thing

tony-- commented on src/main/java/org/mandas/docker/client/LogReader.java:
```diff
@@ -66,15 +63,31 @@ public LogMessage nextMessage() throws IOException {
 
+  private int readAllBytes(byte[] buf) throws IOException {
+	int total = 0;
+    while (total < buf.length) {
```
Do you need this while loop and the 3 parameter version of `read`?  Reading the description [here](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[])) it seems like this will be functionally equivalent to `stream.read(buf);`

dmandalidis replied:
That was my initial impression as well but tests failed at the end. The difference between the `stream.read(buf)` and this while loop is that the first one could return if just a single byte is read. What we need at this point is to block until we reach the buffer size or EOF. The previous `ByteStreams.readFully` was doing this internally as well

tony-- commented on src/main/java/org/mandas/docker/client/auth/gcr/ContainerRegistryAuthSupplier.java:
```diff
@@ -127,7 +131,7 @@ public static Builder forCredentials(final GoogleCredentials credentials) {
 
     private Collection<String> scopes =
-        ImmutableList.of("https://www.googleapis.com/auth/devstorage.read_write");
+        Collections.singletonList("https://www.googleapis.com/auth/devstorage.read_write");
```
Nice, I had no idea this existed.

tony-- commented on src/main/java/org/mandas/docker/client/messages/HostConfig.java:
```diff
@@ -314,8 +313,9 @@ public static Builder builder() {
 	    for (final String extraHost : extraHosts()) {
-	      checkArgument(extraHost.contains(":"),
-	      "extra host arg '%s' must contain a ':'", extraHost);
+	      if (!extraHost.contains(".")) {
```
"." should be ":" here, I think?

dmandalidis replied:
good catch

tony-- commented on src/main/java/org/mandas/docker/client/messages/RegistryConfigs.java:
```diff
@@ -66,22 +68,18 @@ public static RegistryConfigs create(final Map<String, RegistryAuth> configs) {
+    		}
+    		return value;
+    	}));
+    
```
nit: mix of 2, 4, 8 indent lengths

tony-- commented on src/test/java/org/mandas/docker/client/DefaultLogStreamTest.java:
```diff
@@ -45,12 +45,13 @@ public void testAttach() throws Exception {
-    logStream.attach(stdout, stderr);
+    try (final ByteArrayOutputStream stdout = new ByteArrayOutputStream();
+    		final ByteArrayOutputStream stderr = new ByteArrayOutputStream()) {
+    	logStream.attach(stdout, stderr);
```
nit: extra indent

Why did you need to add the try-with-resources here?

dmandalidis replied:
Didn't resist the temptation to close the byte array streams (resource leaks even in tests can cause problems as well)

tony-- reviewed (changes requested):
I'm done reviewing, please take a look at my comments.  Thanks for all the effort here.

dmandalidis commented:
@tony-- thanks a million for the effort reviewing this. I think I 've addressed your comments, please take a final look when you feel so.

As a side note, I think I 've been misusing the `Resolve conversation` button, I believe you should do that if you see that each one of your comments is addressed. Me doing this seems a bit rude, sorry :)

tony-- reviewed (approved):
LGTM
